File: electron-app/electron/playback/skip-segments.js
const ANISKIP_BASE_URL = "https://api.aniskip.com/v2/skip-times";
const THEINTRODB_BASE_URL = "https://api.theintrodb.org/v3/media";
const DEFAULT_TIMEOUT_MS = 3500;
const CACHE_TTL_MS = 24 * 60 * 60 * 1000;

const responseCache = new Map();

const TYPE_PRIORITY = {
  recap: 0,
  intro: 1,
  credits: 2,
};

const PLAYBACK_SKIP_SEGMENTS = [
  ["recap", "recap"],
  ["intro", "intro"],
  ["credits", "credits"],
];

function finiteNumber(value) {
  if (value === null || value === undefined || value === "") return null;
  const number = Number(value);
  return Number.isFinite(number) ? number : null;
}

function normalizeSegment(type, startValue, endValue, durationSeconds, provider) {
  const duration = finiteNumber(durationSeconds);
  const start = finiteNumber(startValue);
  const rawEnd = endValue === null || endValue === undefined ? duration : finiteNumber(endValue);
  if (start === null || rawEnd === null || rawEnd <= start) return null;

  const clampedStart = Math.max(0, start);
  const clampedEnd = duration && duration > 0
    ? Math.min(duration, rawEnd)
    : rawEnd;
  if (clampedEnd - clampedStart < 3) return null;

  return {
    type,
    start: clampedStart,
    end: clampedEnd,
    provider,
  };
}

function dedupeAndSort(segments) {
  const unique = new Map();
  for (const segment of segments.filter(Boolean)) {
    const key = `${segment.type}:${Math.round(segment.start * 4)}:${Math.round(segment.end * 4)}`;
    if (!unique.has(key)) unique.set(key, segment);
  }
  return [...unique.values()].sort((left, right) => (
    left.start - right.start
    || TYPE_PRIORITY[left.type] - TYPE_PRIORITY[right.type]
    || left.end - right.end
  ));
}

export function normalizeAniSkipResponse(payload, durationSeconds) {
  if (!payload?.found || !Array.isArray(payload.results)) return [];

  const typeMap = {
    op: "intro",
    "mixed-op": "intro",
    recap: "recap",
    ed: "credits",
    "mixed-ed": "credits",
  };

  return dedupeAndSort(payload.results.map((result) => {
    const type = typeMap[String(result?.skipType || "").toLowerCase()];
    if (!type) return null;

    const sourceDuration = finiteNumber(result?.episodeLength);
    const actualDuration = finiteNumber(durationSeconds);
    const durationDifference = sourceDuration && actualDuration
      ? actualDuration - sourceDuration
      : 0;
    // AniSkip timestamps are relative to the episode duration used when the
    // segment was submitted. Small runtime differences usually come from
    // alternate releases with a longer/shorter pre-roll; large differences
    // indicate a mismatched file and should not shift the segment wholesale.
    const offset = Math.abs(durationDifference) <= 120 ? durationDifference : 0;
    const intervalStart = finiteNumber(result?.interval?.startTime);
    const intervalEnd = finiteNumber(result?.interval?.endTime);
    if (intervalStart === null || intervalEnd === null) return null;
    return normalizeSegment(
      type,
      intervalStart + offset,
      intervalEnd + offset,
      actualDuration,
      "aniskip",
    );
  }));
}

export function normalizeTheIntroDBResponse(payload, durationSeconds) {
  if (!payload || typeof payload !== "object") return [];

  const segments = [];
  for (const [responseKey, type] of PLAYBACK_SKIP_SEGMENTS) {
    const rows = Array.isArray(payload[responseKey]) ? payload[responseKey] : [];
    for (const row of rows) {
      const startMs = finiteNumber(row?.start_ms);
      const endMs = finiteNumber(row?.end_ms);
      if (startMs === null) continue;
      segments.push(normalizeSegment(
        type,
        startMs / 1000,
        endMs === null ? null : endMs / 1000,
        durationSeconds,
        "theintrodb",
      ));
    }
  }
  return dedupeAndSort(segments);
}

async function fetchJson(url, { fetchImpl, signal, timeoutMs }) {
  const controller = new AbortController();
  const abort = () => controller.abort();
  if (signal?.aborted) controller.abort();
  else signal?.addEventListener("abort", abort, { once: true });
  const timeout = setTimeout(() => controller.abort(), timeoutMs);

  try {
    const response = await fetchImpl(url, {
      signal: controller.signal,
      headers: { Accept: "application/json" },
    });
    if (!response.ok) throw new Error(`skip segment request returned ${response.status}`);
    return await response.json();
  } finally {
    clearTimeout(timeout);
    signal?.removeEventListener("abort", abort);
  }
}

function cacheKeyFor(context) {
  const duration = Math.round(finiteNumber(context.durationSeconds) || 0);
  if (context.kind === "anime") {
    return `aniskip:${context.malId}:${context.episode}:${duration}`;
  }
  return `theintrodb:${context.tmdbId || context.imdbId}:${context.season}:${context.episode}:${duration}`;
}

function readCache(key) {
  const cached = responseCache.get(key);
  if (!cached) return null;
  if (Date.now() - cached.savedAt > CACHE_TTL_MS) {
    responseCache.delete(key);
    return null;
  }
  return cached.segments;
}

function writeCache(key, segments) {
  responseCache.set(key, { savedAt: Date.now(), segments });
  if (responseCache.size <= 250) return;
  const oldestKey = responseCache.keys().next().value;
  if (oldestKey) responseCache.delete(oldestKey);
}

export async function getSkipSegments(context, options = {}) {
  const fetchImpl = options.fetchImpl || globalThis.fetch;
  if (typeof fetchImpl !== "function") return [];

  const durationSeconds = finiteNumber(context?.durationSeconds);
  const kind = String(context?.kind || "").toLowerCase();
  const season = Number(context?.season);
  const episode = Number(kind === "anime"
    ? context?.absoluteEpisode || context?.episode
    : context?.episode);
  if (!durationSeconds || durationSeconds <= 0 || !Number.isInteger(episode) || episode <= 0) return [];

  const timeoutMs = Number(options.timeoutMs) > 0 ? Number(options.timeoutMs) : DEFAULT_TIMEOUT_MS;
  const normalizedContext = {
    ...context,
    kind,
    durationSeconds,
    season,
    episode,
  };
  const key = cacheKeyFor(normalizedContext);
  const cached = readCache(key);
  if (cached) return cached;

  let url;
  let normalize;
  if (kind === "anime") {
    const malId = Number(context?.malId);
    if (!Number.isInteger(malId) || malId <= 0) return [];
    url = new URL(`${ANISKIP_BASE_URL}/${malId}/${episode}`);
    for (const type of ["op", "ed", "recap"]) {
      // AniSkip expects repeated `types` parameters. Its server currently
      // returns HTTP 500 for the otherwise-common `types[]` array spelling.
      url.searchParams.append("types", type);
    }
    url.searchParams.set("episodeLength", String(Math.round(durationSeconds)));
    normalize = normalizeAniSkipResponse;
  } else if (kind === "tv") {
    const tmdbId = Number(context?.tmdbId);
    const imdbId = String(context?.imdbId || "").trim();
    if ((!Number.isInteger(tmdbId) || tmdbId <= 0) && !/^tt\d{7,10}$/i.test(imdbId)) return [];
    if (!Number.isInteger(season) || season < 0) return [];
    url = new URL(THEINTRODB_BASE_URL);
    if (Number.isInteger(tmdbId) && tmdbId > 0) url.searchParams.set("tmdb_id", String(tmdbId));
    else url.searchParams.set("imdb_id", imdbId);
    url.searchParams.set("season", String(season));
    url.searchParams.set("episode", String(episode));
    url.searchParams.set("duration_ms", String(Math.round(durationSeconds * 1000)));
    normalize = normalizeTheIntroDBResponse;
  } else {
    return [];
  }

  try {
    const fetchWithAnimeDurationFallback = async (requestUrl) => {
      try {
        return await fetchJson(requestUrl, {
          fetchImpl,
          signal: options.signal,
          timeoutMs,
        });
      } catch (error) {
        if (kind !== "anime" || options.signal?.aborted) throw error;

        // AniSkip returns HTTP 500 when no entry closely matches some supplied
        // runtimes. A duration-agnostic retry returns the available timestamps;
        // normalization below still clamps and adjusts them to the real file.
        const fallbackUrl = new URL(requestUrl);
        fallbackUrl.searchParams.set("episodeLength", "0");
        return fetchJson(fallbackUrl, {
          fetchImpl,
          signal: options.signal,
          timeoutMs,
        });
      }
    };

    const payload = await fetchWithAnimeDurationFallback(url);
    let segments = normalize(payload, durationSeconds);

    if (kind === "anime") {
      const missingTypes = new Set(["intro", "credits"]);
      for (const segment of segments) missingTypes.delete(segment.type);
      if (missingTypes.size > 0) {
        // AniSkip fails when standard and mixed filters are combined in one
        // request. Query mixed OP/ED only when their standard counterpart is
        // absent, then use them strictly to fill those gaps.
        const mixedUrl = new URL(`${ANISKIP_BASE_URL}/${normalizedContext.malId}/${episode}`);
        mixedUrl.searchParams.append("types", "mixed-op");
        mixedUrl.searchParams.append("types", "mixed-ed");
        mixedUrl.searchParams.set("episodeLength", String(Math.round(durationSeconds)));
        const mixedPayload = await fetchWithAnimeDurationFallback(mixedUrl);
        const mixedSegments = normalizeAniSkipResponse(mixedPayload, durationSeconds)
          .filter((segment) => missingTypes.has(segment.type));
        segments = dedupeAndSort([...segments, ...mixedSegments]);
      }
    }
    writeCache(key, segments);
    return segments;
  } catch (error) {
    if (error?.name !== "AbortError") {
      console.warn(`[SkipSegments] ${kind} lookup unavailable:`, error?.message || error);
    }
    return [];
  }
}

export function clearSkipSegmentCache() {
  responseCache.clear();
}

File: torrent-streamer/internal/skipsegments/skipsegments.go
// Package skipsegments resolves intro/recap/credits timestamps for the
// mobile player (skip-intro support where timestamps exist). It is the Go
// port of the desktop Electron helper (electron/playback/skip-segments.js):
// AniSkip for anime, TheIntroDB for TV, normalized to the actual file
// duration and cached for 24 h.
//
// The endpoint lives on the server (not the client) because both provider
// APIs are third-party origins with unknown CORS posture, and the response
// cache should be shared per household, not per device.
//
// GET /skip-segments?kind=tv|anime&tmdbId=&imdbId=&malId=&season=&episode=
//
//	&absoluteEpisode=&durationSeconds=
package skipsegments

import (
	"context"
	"encoding/json"
	"io"
	"math"
	"net/http"
	"net/url"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"sync"
	"time"

	"torrent-streamer/internal/middleware"
)

const (
	aniskipBaseURL   = "https://api.aniskip.com/v2/skip-times"
	theintrodbBase   = "https://api.theintrodb.org/v3/media"
	requestTimeout   = 3500 * time.Millisecond
	cacheTTL         = 24 * time.Hour
	cacheMaxEntries  = 250
	minSegmentSecs   = 3
	durationSlackSec = 120
)

// Segment is one normalized, clamped skip window (seconds).
type Segment struct {
	Type     string  `json:"type"` // intro | recap | credits
	Start    float64 `json:"start"`
	End      float64 `json:"end"`
	Provider string  `json:"provider"`
}

var typePriority = map[string]int{"recap": 0, "intro": 1, "credits": 2}

type cacheEntry struct {
	savedAt  time.Time
	segments []Segment
}

var (
	cacheMu sync.Mutex
	cache   = map[string]cacheEntry{}
)

func finite(v interface{}) float64 {
	switch n := v.(type) {
	case float64:
		if !isNaN(n) {
			return n
		}
	case int:
		return float64(n)
	case string:
		if f, err := strconv.ParseFloat(strings.TrimSpace(n), 64); err == nil && !isNaN(f) {
			return f
		}
	}
	return math.NaN()
}

func isNaN(f float64) bool { return f != f }

func normalizeSegment(kind string, start, end, duration float64, provider string) *Segment {
	if isNaN(start) {
		return nil
	}
	if isNaN(end) && duration > 0 {
		end = duration
	}
	if isNaN(end) || end <= start {
		return nil
	}
	if start < 0 {
		start = 0
	}
	if duration > 0 && end > duration {
		end = duration
	}
	if end-start < minSegmentSecs {
		return nil
	}
	return &Segment{Type: kind, Start: start, End: end, Provider: provider}
}

func dedupeAndSort(segments []Segment) []Segment {
	seen := map[string]bool{}
	out := make([]Segment, 0, len(segments))
	for _, s := range segments {
		key := s.Type + ":" + strconv.Itoa(int(math.Round(s.Start*4))) + ":" + strconv.Itoa(int(math.Round(s.End*4)))
		if seen[key] {
			continue
		}
		seen[key] = true
		out = append(out, s)
	}
	sort.SliceStable(out, func(i, j int) bool {
		if out[i].Start != out[j].Start {
			return out[i].Start < out[j].Start
		}
		if typePriority[out[i].Type] != typePriority[out[j].Type] {
			return typePriority[out[i].Type] < typePriority[out[j].Type]
		}
		return out[i].End < out[j].End
	})
	if out == nil {
		out = []Segment{}
	}
	return out
}

func fetchJSON(ctx context.Context, requestURL string) (map[string]interface{}, error) {
	ctx, cancel := context.WithTimeout(ctx, requestTimeout)
	defer cancel()
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, requestURL, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Accept", "application/json")
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, errStatus(resp.StatusCode)
	}
	var payload map[string]interface{}
	if err := json.NewDecoder(io.LimitReader(resp.Body, 1<<20)).Decode(&payload); err != nil {
		return nil, err
	}
	return payload, nil
}

type statusError int

func (e statusError) Error() string { return "skip segment request returned " + strconv.Itoa(int(e)) }
func errStatus(code int) error      { return statusError(code) }

func asArray(v interface{}) []interface{} {
	if list, ok := v.([]interface{}); ok {
		return list
	}
	return nil
}

func asObject(v interface{}) map[string]interface{} {
	if obj, ok := v.(map[string]interface{}); ok {
		return obj
	}
	return nil
}

func normalizeAniSkip(payload map[string]interface{}, duration float64) []Segment {
	if payload == nil || payload["found"] != true {
		return nil
	}
	typeMap := map[string]string{
		"op":       "intro",
		"mixed-op": "intro",
		"recap":    "recap",
		"ed":       "credits",
		"mixed-ed": "credits",
	}
	var segments []Segment
	for _, raw := range asArray(payload["results"]) {
		result := asObject(raw)
		if result == nil {
			continue
		}
		kind, ok := typeMap[strings.ToLower(stringValue(result["skipType"]))]
		if !ok {
			continue
		}
		sourceDuration := finite(result["episodeLength"])
		offset := 0.0
		// AniSkip timestamps are relative to the submitted episode runtime.
		// Small differences usually come from alternate releases with a
		// longer/shorter pre-roll; large ones mean a mismatched file.
		if !isNaN(sourceDuration) && duration > 0 && math.Abs(duration-sourceDuration) <= durationSlackSec {
			offset = duration - sourceDuration
		}
		interval := asObject(result["interval"])
		if interval == nil {
			continue
		}
		start := finite(interval["startTime"])
		end := finite(interval["endTime"])
		if isNaN(start) || isNaN(end) {
			continue
		}
		if seg := normalizeSegment(kind, start+offset, end+offset, duration, "aniskip"); seg != nil {
			segments = append(segments, *seg)
		}
	}
	return segments
}

func stringValue(v interface{}) string {
	if s, ok := v.(string); ok {
		return s
	}
	return ""
}

var playbackSkipSegments = [][2]string{{"recap", "recap"}, {"intro", "intro"}, {"credits", "credits"}}

func normalizeTheIntroDB(payload map[string]interface{}, duration float64) []Segment {
	if payload == nil {
		return nil
	}
	var segments []Segment
	for _, pair := range playbackSkipSegments {
		for _, raw := range asArray(payload[pair[0]]) {
			row := asObject(raw)
			if row == nil {
				continue
			}
			startMs := finite(row["start_ms"])
			if isNaN(startMs) {
				continue
			}
			endMs := finite(row["end_ms"])
			end := math.NaN() // missing end = open-ended (duration clamp applies)
			if !isNaN(endMs) {
				end = endMs / 1000
			}
			if seg := normalizeSegment(pair[1], startMs/1000, end, duration, "theintrodb"); seg != nil {
				segments = append(segments, *seg)
			}
		}
	}
	return segments
}

func cacheKey(kind string, context url.Values, duration float64, season, episode int) string {
	switch kind {
	case "anime":
		return "aniskip:" + context.Get("malId") + ":" + strconv.Itoa(episode) + ":" + strconv.Itoa(int(math.Round(duration)))
	default:
		id := context.Get("tmdbId")
		if id == "" {
			id = context.Get("imdbId")
		}
		return "theintrodb:" + id + ":" + strconv.Itoa(season) + ":" + strconv.Itoa(episode) + ":" + strconv.Itoa(int(math.Round(duration)))
	}
}

func readCache(key string) []Segment {
	cacheMu.Lock()
	defer cacheMu.Unlock()
	entry, ok := cache[key]
	if !ok || time.Since(entry.savedAt) > cacheTTL {
		delete(cache, key)
		return nil
	}
	return entry.segments
}

func writeCache(key string, segments []Segment) {
	cacheMu.Lock()
	defer cacheMu.Unlock()
	cache[key] = cacheEntry{savedAt: time.Now(), segments: segments}
	if len(cache) <= cacheMaxEntries {
		return
	}
	// Bounded memory: drop the oldest entry (insertion order via scan).
	var oldestKey string
	var oldest time.Time
	for k, v := range cache {
		if oldestKey == "" || v.savedAt.Before(oldest) {
			oldestKey, oldest = k, v.savedAt
		}
	}
	delete(cache, oldestKey)
}

// Handler serves GET /skip-segments.
func Handler(w http.ResponseWriter, r *http.Request) {
	middleware.EnableCORS(w)
	if r.Method == http.MethodOptions {
		return
	}
	if r.Method != http.MethodGet {
		http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
		return
	}
	q := r.URL.Query()
	kind := strings.ToLower(q.Get("kind"))
	duration := finite(q.Get("durationSeconds"))
	season, _ := strconv.Atoi(q.Get("season"))
	episode, _ := strconv.Atoi(q.Get("episode"))
	if kind == "anime" {
		if abs, err := strconv.Atoi(q.Get("absoluteEpisode")); err == nil && abs > 0 {
			episode = abs
		}
	}
	if math.IsNaN(duration) || math.IsInf(duration, 0) || duration <= 0 || duration > 7*24*3600 || episode <= 0 {
		writeSegments(w, []Segment{})
		return
	}

	key := cacheKey(kind, q, duration, season, episode)
	if cached := readCache(key); cached != nil {
		writeSegments(w, cached)
		return
	}

	ctx := r.Context()
	var segments []Segment
	switch kind {
	case "anime":
		malID, _ := strconv.Atoi(q.Get("malId"))
		if malID <= 0 {
			writeSegments(w, []Segment{})
			return
		}
		base := aniskipBaseURL + "/" + strconv.Itoa(malID) + "/" + strconv.Itoa(episode)
		length := strconv.Itoa(int(math.Round(duration)))
		payload, err := fetchJSON(ctx, base+"?types=op&types=ed&types=recap&episodeLength="+length)
		if err != nil {
			// AniSkip returns HTTP 500 when no entry closely matches the
			// supplied runtime; retry duration-agnostically (normalization
			// below still clamps and offsets to the real file).
			payload, err = fetchJSON(ctx, base+"?types=op&types=ed&types=recap&episodeLength=0")
		}
		if err == nil {
			segments = normalizeAniSkip(payload, duration)
			missing := map[string]bool{"intro": true, "credits": true}
			for _, s := range segments {
				delete(missing, s.Type)
			}
			if len(missing) > 0 {
				// AniSkip fails when standard and mixed filters are combined;
				// query mixed OP/ED only to fill the gaps.
				if mixed, err := fetchJSON(ctx, base+"?types=mixed-op&types=mixed-ed&episodeLength="+length); err == nil {
					for _, s := range normalizeAniSkip(mixed, duration) {
						if missing[s.Type] {
							segments = append(segments, s)
						}
					}
					segments = dedupeAndSort(segments)
				}
			}
		}
	case "tv":
		tmdbID, _ := strconv.Atoi(q.Get("tmdbId"))
		imdbID := strings.TrimSpace(q.Get("imdbId"))
		if tmdbID <= 0 && !imdbIDPattern.MatchString(imdbID) {
			writeSegments(w, []Segment{})
			return
		}
		if season < 0 {
			season = 0
		}
		params := url.Values{}
		if tmdbID > 0 {
			params.Set("tmdb_id", strconv.Itoa(tmdbID))
		} else {
			params.Set("imdb_id", imdbID)
		}
		params.Set("season", strconv.Itoa(season))
		params.Set("episode", strconv.Itoa(episode))
		params.Set("duration_ms", strconv.Itoa(int(math.Round(duration*1000))))
		if payload, err := fetchJSON(ctx, theintrodbBase+"?"+params.Encode()); err == nil {
			segments = normalizeTheIntroDB(payload, duration)
		}
	default:
		writeSegments(w, []Segment{})
		return
	}

	if segments == nil {
		segments = []Segment{}
	}
	writeCache(key, segments)
	writeSegments(w, segments)
}

var imdbIDPattern = regexp.MustCompile(`^tt\d{7,10}$`)

func writeSegments(w http.ResponseWriter, segments []Segment) {
	w.Header().Set("Content-Type", "application/json")
	w.Header().Set("Cache-Control", "no-store")
	_ = json.NewEncoder(w).Encode(map[string]interface{}{"segments": segments})
}
